### ISSUE TITLE
SyncServicerContext context.abort allows code to run after the abort statement is called

### ISSUE THREAD

**Issue by erikhartigltn** [grpc/grpc#37518]
### What version of gRPC and what language are you using?
grpc_version: 1.65.4
grpc_tools_version: 1.65.4
language: Python

### What operating system (Linux, Windows,...) and version?
Originally noticed on AlmaLinux 8 containers running on a CentOS box
Replicated on MacOS Sonoma 14.5

### What runtime / compiler are you using (e.g. python version or version of gcc)
python version: 3.12.4

### What did you do?
Please provide either 1) A unit test for reproducing the bug or 2) Specific steps for us to follow to reproduce the bug. If there’s not enough information to debug the problem, gRPC team may close the issue at their discretion. You’re welcome to re-open the issue once you have a reproduction.
Below is a zip file including Python code and a readme to reproduce the issue.
[pythonProject.zip](https://github.com/user-attachments/files/16639909/pythonProject.zip)

### What did you expect to see?
I expected when context.abort() was called in a synchronous rpc when using an asyncio gprc server for the server to not run any of the code after the context.abort().

### What did you see instead?
Code executing after the context.abort() despite the client receiving the message generated by the abort.

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

### Anything else we should know about your project / environment?
We were in the process of converting from using a synchronous grpc server to an asyncio server.  However, our backend is quite large (we have 100s of endpoints some of which are quite complicated) and we planned to convert over endpoints as necessary slowly.  However, after we started testing, we noticed that some of our tests, which checked that users met permissions to be allowed to create objects, were causing objects to be created.  After investigation, we determined that all our code after our context.abort statements was running despite us getting back an abort response from the server.  This is a huge problem and prevents us from moving forward with using asyncio in our grpc server.

**Comment by sreenithi** [comment]
Hi @erikhartigltn, thank you for reporting the issue and providing good code to reproduce it. I could see the output you explained when I ran the provided code. However I was able to solve it when I converted the call handler method in server to an async function and using `await context.abort` as below:
```
async def GetFeature(self, request: route_guide_pb2.Point, context):
    await context.abort("here")
    for i in range(100):
        print(time.time())
 ```
 Running the code above, should solve your issue. Do try it out and let us know if the issue persists.

**Comment by erikhartigltn** [comment]
Yes that is true that the code you post does work as expected.  However that isn't really relevant to the bug I am reporting.  First of all the type of the context object is different for this function then the one I sent
`class RouteGuideServicer(route_guide_pb2_grpc.RouteGuideServicer):
    async def GetFeature(self, request: route_guide_pb2.Point, context):
        print(type(context))
`
yields <class 'grpc._cython.cygrpc._ServicerContext'>
where as 
`class RouteGuideServicer(route_guide_pb2_grpc.RouteGuideServicer):
    def GetFeature(self, request: route_guide_pb2.Point, context):
        print(type(context))
        context.abort("here")
`
yields <class 'grpc._cython.cygrpc._SyncServicerContext'>.

Additionally the grpc documentation https://grpc.github.io/grpc/python/grpc_asyncio.html#create-server specifically the part talking about the migration thread pool says it is:
A futures.ThreadPoolExecutor to be used by the Server to execute non-AsyncIO RPC handlers for migration purpose.

This suggests that with an aio server I should still be able to use existing non-async endpoints for purposes of migration.  Migration is what we are trying to do, we have hundreds of endpoints and tens of thousands of lines of code so changing everything to async and testing it all at once would be a huge undertaking.  So from me reading of the documentation I would not expect to have to change the code to async for this behavior not to show up.

**Comment by erikhartigltn** [comment]
I was just checking in to see if my response contains all the information it needs to.  This is still an issue and the initial response acknowledges the problem and seems to state that the asyncio migration feature is not really supported.  I want to know if this will get fixed at some point or if I should just treat the migration feature as though it doesn't exist and accept we would have to convert all our endpoints at once (probably never going to happen) or just not use asyncio.

**Comment by sreenithi** [comment]
hi, I'm sorry, I didn't have the chance to come back to this issue. It looks like a bug when using sync handlers with an async server. We would like to fix this issue, but not sure on the exact timeline as of now. I'll get back once I discuss the feasibility with our team. But meanwhile, we are also open to any external contributions.

**Comment by mohitdebian** [comment]
🚨 **Fix Submitted!** 🚨

Hey everyone! I noticed this critical bug has been open for a while and the previous PR (#39136) stalled out over a year ago. 

I've submitted a fresh PR here: **#43126** which implements the same thread-safe `future.result()` fix along with robust regression tests. Hopefully we can get this merged soon so users can safely migrate to the asyncio server without endpoints silently continuing execution after an abort! 🚀

**Comment by mohitdebian** [comment]
I opened PR #43126 for this issue. It fixes the bug where `context.abort()` in a synchronous handler on an asyncio server does not halt execution by properly raising an exception.

**Comment by mohitdebian** [comment]
I opened PR #43126 for this issue. It fixes the bug where `context.abort()` in a synchronous handler on an asyncio server does not halt execution by properly raising an exception.

**Comment by LubabaKhalid** [comment]
I'd like to work on this task, Can you please assign it to me?
